File: java/src/com/nttdocomo/device/IrRemoteControl.java
package com.nttdocomo.device;

import java.util.Arrays;
import java.util.concurrent.atomic.AtomicBoolean;

public class IrRemoteControl {

    /**
     * Indicates the code output pattern is High-first (= 0).
     */
    public static final int PATTERN_HL = 0;
    /**
     * Indicates the code output pattern is Low-first (= 1).
     */
    public static final int PATTERN_LH = 1;

    private static final IrRemoteControl INSTANCE = new IrRemoteControl();

    public static IrRemoteControl getIrRemoteControl() {
        return INSTANCE;
    }

    private int carrierHighUs = -1;
    private int carrierLowUs = -1;

    private Pulse code0;
    private Pulse code1;

    private Thread sendThread;
    private final AtomicBoolean sending = new AtomicBoolean(false);

    protected IrRemoteControl() {
    }

    /**
     * @param highDuration carrier High duration (0.1 microsecond units).
     * @param lowDuration  carrier Low duration (0.1 microsecond units).
     */
    public synchronized void setCarrier(int highDuration, int lowDuration) {
        if (highDuration <= 0 || lowDuration <= 0) {
            throw new IllegalArgumentException("Carrier durations must be > 0");
        }
        this.carrierHighUs = highDuration;
        this.carrierLowUs = lowDuration;

        log("Carrier set: HIGH=%dus LOW=%dus", highDuration, lowDuration);
    }

    /**
     * Sets the logical "0" pulse information.
     *
     * @param pattern      PATTERN_HL or PATTERN_LH
     * @param highDuration pulse High duration
     * @param lowDuration  pulse Low duration
     */
    public synchronized void setCode0(int pattern, int highDuration, int lowDuration) {
        this.code0 = new Pulse(pattern, highDuration, lowDuration);
        log("Code0 set: %s", code0);
    }

    /**
     * Sets the logical "1" pulse information.
     *
     * @param pattern      PATTERN_HL or PATTERN_LH
     * @param highDuration pulse High duration
     * @param lowDuration  pulse Low duration
     */
    public synchronized void setCode1(int pattern, int highDuration, int lowDuration) {
        this.code1 = new Pulse(pattern, highDuration, lowDuration);
        log("Code1 set: %s", code1);
    }

    /**
     * @param numFrames number of frames to send (must be >= 1).
     * @param frames    array of frames to send.
     */
    public void send(int numFrames, IrRemoteControlFrame[] frames) {
        send(numFrames, frames, 10, Integer.MAX_VALUE);
    }

    /**
     * @param numFrames number of frames to send (must be >= 1).
     * @param frames    array of frames to send.
     * @param timeout   timeout value (seconds).
     */
    public void send(int numFrames, IrRemoteControlFrame[] frames, int timeout) {
        send(numFrames, frames, timeout, Integer.MAX_VALUE);
    }

    /**
     * @param numFrames number of frames to send (must be >= 1).
     * @param frames    array of frames to send.
     * @param timeout   timeout value (seconds).
     * @param count     number of times to transmit the signal.
     */
    public synchronized void send(int numFrames,
                                  IrRemoteControlFrame[] frames,
                                  int timeout,
                                  int count) {

        validateSendArgs(numFrames, frames, timeout, count);
        stop();
        sending.set(true);

        sendThread = new Thread(() -> runSendLoop(numFrames, frames, timeout, count),
                "IrRemoteControl-Send");
        sendThread.setDaemon(true);
        sendThread.start();
    }

    private void runSendLoop(int numFrames,
                             IrRemoteControlFrame[] frames,
                             int timeoutSec,
                             int count) {

        log("=== IR SEND START ===");

        long endTimeMs = System.currentTimeMillis() + timeoutSec * 1000L;

        try {
            for (int cycle = 1; cycle <= count; cycle++) {
                if (shouldStop(endTimeMs)) break;
                log("Send cycle %d", cycle);
                boolean completed = sendOneCycle(numFrames, frames, endTimeMs);
                if (!completed) break;
            }
        } finally {
            sending.set(false);
            log("=== IR SEND END ===");
        }
    }

    private boolean sendOneCycle(int numFrames,
                                 IrRemoteControlFrame[] frames,
                                 long endTimeMs) {

        for (int i = 0; i < numFrames; i++) {
            if (shouldStop(endTimeMs)) return false;

            IrRemoteControlFrame f = frames[i];
            logFrame(i, f);

            boolean shouldContinue = sendFrameWithRepeats(i, f, endTimeMs);
            if (!shouldContinue) return false;
        }
        return true;
    }

    private boolean sendFrameWithRepeats(int index,
                                         IrRemoteControlFrame f,
                                         long endTimeMs) {

        int repeatCount = f.getRepeatCount();

        if (repeatCount == IrRemoteControlFrame.COUNT_INFINITE) {
            log("Frame %d repeats infinitely", index);

            while (!shouldStop(endTimeMs)) {
                sleepFrame(f);
            }

            return false;
        } else {
            for (int i = 0; i < repeatCount; i++) {
                if (shouldStop(endTimeMs)) return false;
                sleepFrame(f);
            }

            return true;
        }
    }

    private boolean shouldStop(long endTimeMs) {
        return !sending.get() || System.currentTimeMillis() >= endTimeMs || Thread.currentThread().isInterrupted();
    }

    private void validateSendArgs(int numFrames,
                                  IrRemoteControlFrame[] frames,
                                  int timeout,
                                  int count) {
        if (frames == null) throw new NullPointerException("frames");
        if (numFrames <= 0) throw new ArrayIndexOutOfBoundsException("numFrames <= 0");
        if (frames.length < numFrames) throw new ArrayIndexOutOfBoundsException("frames too short");
        if (timeout <= 0) throw new IllegalArgumentException("timeout <= 0");
        if (count <= 0) throw new IllegalArgumentException("count <= 0");

        if (carrierHighUs <= 0 || carrierLowUs <= 0) {
            throw new IllegalArgumentException("Carrier not configured");
        }
        if (code0 == null || code1 == null) {
            throw new IllegalArgumentException("Code pulses not configured");
        }

        for (int i = 0; i < numFrames; i++) {
            IrRemoteControlFrame f = frames[i];
            if (f == null) throw new NullPointerException("frame[" + i + "] is null");
            f.validate();
        }
    }


    public synchronized void stop() {
        if (!sending.get()) return;

        sending.set(false);

        if (sendThread != null) {
            try {
                sendThread.join(200);
            } catch (InterruptedException ignored) {
                Thread.currentThread().interrupt();
            }
            sendThread = null;
        }

        log("IR transmission stopped");
    }

    private void sleepFrame(IrRemoteControlFrame f) {
        try {
            long ms = f.getFrameDuration() / 10; // 0.1 ms units --> ms
            if (ms > 0) Thread.sleep(ms);
        } catch (InterruptedException ignored) {
            Thread.currentThread().interrupt();
        }
    }

    private void logFrame(int index, IrRemoteControlFrame frame) {
        log(
                "Frame[%d]: bits=%d data=%s start(H=%dus L=%dus) stop(H=%dus) repeat=%d dur=%d(0.1ms)",
                index,
                frame.getDataBitLength(),
                Arrays.toString(frame.getDataBytes()),
                frame.getStartHighUs(),
                frame.getStartLowUs(),
                frame.getStopHighUs(),
                frame.getRepeatCount(),
                frame.getFrameDuration()
        );
    }

    private static void log(String fmt, Object... args) {
        System.out.println("[IR] " + String.format(fmt, args));
    }

    private static final class Pulse {
        final int pattern;
        final int highDuration;
        final int lowDuration;

        Pulse(int pattern, int highDuration, int lowDuration) {
            if (pattern != PATTERN_HL && pattern != PATTERN_LH) {
                throw new IllegalArgumentException("Invalid pattern: " + pattern);
            }
            if (highDuration <= 0 || lowDuration <= 0) {
                throw new IllegalArgumentException("Pulse durations must be > 0");
            }
            this.pattern = pattern;
            this.highDuration = highDuration;
            this.lowDuration = lowDuration;
        }

        @Override
        public String toString() {
            return (pattern == PATTERN_HL ? "HL" : "LH") +
                    " HIGH=" + highDuration + "us LOW=" + lowDuration + "us";
        }
    }
}

File: web/src/com/nttdocomo/device/IrRemoteControl.ts
import { IrRemoteControlFrame } from './IrRemoteControlFrame';

export class IrRemoteControl {
	/**
	 * Indicates the code output pattern is High-first (= 0).
	 */
	public static readonly PATTERN_HL = 0;
	/**
	 * Indicates the code output pattern is Low-first (= 1).
	 */
	public static readonly PATTERN_LH = 1;

	private static readonly INSTANCE = new IrRemoteControl();

	public static getIrRemoteControl(): IrRemoteControl {
		return IrRemoteControl.INSTANCE;
	}

	private carrierHighUs = -1;
	private carrierLowUs = -1;

	private code0: Pulse | null = null;
	private code1: Pulse | null = null;

	// "thread" replacement
	private sending = false;
	private sendToken = 0; // increment to cancel any in-flight loop

	protected constructor() {}

	/**
	 * @param highDuration carrier High duration (0.1 microsecond units).
	 * @param lowDuration  carrier Low duration (0.1 microsecond units).
	 */
	public setCarrier(highDuration: number, lowDuration: number): void {
		if (highDuration <= 0 || lowDuration <= 0) {
			throw new Error('Carrier durations must be > 0');
		}
		this.carrierHighUs = highDuration | 0;
		this.carrierLowUs = lowDuration | 0;

		IrRemoteControl.log(
			'Carrier set: HIGH=%dus LOW=%dus',
			highDuration,
			lowDuration
		);
	}

	/**
	 * Sets the logical "0" pulse information.
	 */
	public setCode0(
		pattern: number,
		highDuration: number,
		lowDuration: number
	): void {
		this.code0 = new Pulse(pattern, highDuration, lowDuration);
		IrRemoteControl.log('Code0 set: %s', this.code0.toString());
	}

	/**
	 * Sets the logical "1" pulse information.
	 */
	public setCode1(
		pattern: number,
		highDuration: number,
		lowDuration: number
	): void {
		this.code1 = new Pulse(pattern, highDuration, lowDuration);
		IrRemoteControl.log('Code1 set: %s', this.code1.toString());
	}

	/**
	 * @param numFrames number of frames to send (must be >= 1).
	 * @param frames    array of frames to send.
	 */
	public send(numFrames: number, frames: IrRemoteControlFrame[]): void;
	/**
	 * @param numFrames number of frames to send (must be >= 1).
	 * @param frames    array of frames to send.
	 * @param timeout   timeout value (seconds).
	 */
	public send(
		numFrames: number,
		frames: IrRemoteControlFrame[],
		timeout: number
	): void;
	/**
	 * @param numFrames number of frames to send (must be >= 1).
	 * @param frames    array of frames to send.
	 * @param timeout   timeout value (seconds).
	 * @param count     number of times to transmit the signal.
	 */
	public send(
		numFrames: number,
		frames: IrRemoteControlFrame[],
		timeout?: number,
		count?: number
	): void {
		const t = timeout ?? 10;
		const c = count ?? Number.MAX_SAFE_INTEGER;
		this.sendInternal(numFrames, frames, t, c);
	}

	private sendInternal(
		numFrames: number,
		frames: IrRemoteControlFrame[],
		timeoutSec: number,
		count: number
	): void {
		this.validateSendArgs(numFrames, frames, timeoutSec, count);

		this.stop();
		this.sending = true;

		const myToken = ++this.sendToken;

		// Fire-and-forget async loop (closest to daemon thread)
		void this.runSendLoop(myToken, numFrames, frames, timeoutSec, count);
	}

	private async runSendLoop(
		token: number,
		numFrames: number,
		frames: IrRemoteControlFrame[],
		timeoutSec: number,
		count: number
	): Promise<void> {
		IrRemoteControl.log('=== IR SEND START ===');

		const endTimeMs = Date.now() + timeoutSec * 1000;

		try {
			for (let cycle = 1; cycle <= count; cycle++) {
				if (this.shouldStop(token, endTimeMs)) break;

				IrRemoteControl.log('Send cycle %d', cycle);

				const completed = await this.sendOneCycle(
					token,
					numFrames,
					frames,
					endTimeMs
				);
				if (!completed) break;
			}
		} finally {
			// Only the latest send loop should turn off sending
			if (this.sendToken === token) this.sending = false;
			IrRemoteControl.log('=== IR SEND END ===');
		}
	}

	private async sendOneCycle(
		token: number,
		numFrames: number,
		frames: IrRemoteControlFrame[],
		endTimeMs: number
	): Promise<boolean> {
		for (let i = 0; i < numFrames; i++) {
			if (this.shouldStop(token, endTimeMs)) return false;

			const f = frames[i];
			this.logFrame(i, f);

			const shouldContinue = await this.sendFrameWithRepeats(
				token,
				i,
				f,
				endTimeMs
			);
			if (!shouldContinue) return false;
		}
		return true;
	}

	private async sendFrameWithRepeats(
		token: number,
		index: number,
		f: IrRemoteControlFrame,
		endTimeMs: number
	): Promise<boolean> {
		const repeatCount = f.getRepeatCount();

		if (repeatCount === IrRemoteControlFrame.COUNT_INFINITE) {
			IrRemoteControl.log('Frame %d repeats infinitely', index);

			while (!this.shouldStop(token, endTimeMs)) {
				await this.sleepFrame(token, f);
			}
			return false;
		} else {
			for (let i = 0; i < repeatCount; i++) {
				if (this.shouldStop(token, endTimeMs)) return false;
				await this.sleepFrame(token, f);
			}
			return true;
		}
	}

	private shouldStop(token: number, endTimeMs: number): boolean {
		return (
			!this.sending || this.sendToken !== token || Date.now() >= endTimeMs
		);
	}

	private validateSendArgs(
		numFrames: number,
		frames: IrRemoteControlFrame[],
		timeout: number,
		count: number
	): void {
		if (frames == null) throw new Error('frames');
		if (numFrames <= 0) throw new Error('numFrames <= 0');
		if (frames.length < numFrames) throw new Error('frames too short');
		if (timeout <= 0) throw new Error('timeout <= 0');
		if (count <= 0) throw new Error('count <= 0');

		if (this.carrierHighUs <= 0 || this.carrierLowUs <= 0) {
			throw new Error('Carrier not configured');
		}
		if (this.code0 == null || this.code1 == null) {
			throw new Error('Code pulses not configured');
		}

		for (let i = 0; i < numFrames; i++) {
			const f = frames[i];
			if (f == null) throw new Error(`frame[${i}] is null`);
			// validate() is protected in my TS frame conversion; call via (f as any) if needed.
			f.validate?.();
		}
	}

	public stop(): void {
		if (!this.sending) return;

		this.sending = false;
		this.sendToken++; // cancels any in-flight loop immediately

		IrRemoteControl.log('IR transmission stopped');
	}

	private async sleepFrame(
		token: number,
		f: IrRemoteControlFrame
	): Promise<void> {
		// 0.1 ms units --> ms (integer)
		const dur01ms = f.getFrameDuration();
		const ms = Math.floor(dur01ms / 10);

		if (ms <= 0) return;
		await IrRemoteControl.sleep(ms);

		// Check cancel quickly after sleeping
		if (this.sendToken !== token) return;
	}

	private logFrame(index: number, frame: IrRemoteControlFrame): void {
		const bytes = frame.getDataBytes();
		IrRemoteControl.log(
			'Frame[%d]: bits=%d data=%s start(H=%dus L=%dus) stop(H=%dus) repeat=%d dur=%d(0.1ms)',
			index,
			frame.getDataBitLength(),
			IrRemoteControl.bytesToString(bytes),
			frame.getStartHighUs(),
			frame.getStartLowUs(),
			frame.getStopHighUs(),
			frame.getRepeatCount(),
			frame.getFrameDuration()
		);
	}

	private static bytesToString(bytes: Uint8Array): string {
		// Similar vibe to Arrays.toString(byte[])
		return `[${Array.from(bytes).join(', ')}]`;
	}

	private static log(fmt: string, ...args: any[]): void {
		// Basic %d / %s support to resemble String.format
		let i = 0;
		const msg = fmt.replace(/%[ds]/g, m => {
			const v = args[i++];
			return m === '%d' ? String(Number(v)) : String(v);
		});
		console.log('[IR] ' + msg, ...args.slice(i));
	}

	private static sleep(ms: number): Promise<void> {
		return new Promise(resolve => setTimeout(resolve, ms));
	}
}

class Pulse {
	public readonly pattern: number;
	public readonly highDuration: number;
	public readonly lowDuration: number;

	constructor(pattern: number, highDuration: number, lowDuration: number) {
		if (
			pattern !== IrRemoteControl.PATTERN_HL &&
			pattern !== IrRemoteControl.PATTERN_LH
		) {
			throw new Error('Invalid pattern: ' + pattern);
		}
		if (highDuration <= 0 || lowDuration <= 0) {
			throw new Error('Pulse durations must be > 0');
		}
		this.pattern = pattern;
		this.highDuration = highDuration | 0;
		this.lowDuration = lowDuration | 0;
	}

	public toString(): string {
		return `${this.pattern === IrRemoteControl.PATTERN_HL ? 'HL' : 'LH'} HIGH=${this.highDuration}us LOW=${this.lowDuration}us`;
	}
}
